### PR TITLE
chore(debugging): add function discovery module

### PR DESCRIPTION
**Pull request by P403n1x87**
This changes adds a new module that can be used to discover all the functions within a module so that they can be retrieved either by line number or by qualified name.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: tests/debugging/function/test_discovery.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        "alias",
+        "Stuff.propertystuff.getter",
+        "Stuff.instancestuff",
+        "Stuff.staticstuff",
+        "Stuff.classstuff",
+        "Stuff.decoratedstuff",
+        "Stuff.doublydecoratedstuff",
+        "Stuff.nestedstuff",
+        "Stuff._Stuff__mangledstuff",
+        "Stuff.__mangledstuff",
```
Comment (majorgreys) [inline tests/debugging/function/test_discovery.py:48]: Should this include `generatorstuff`?

**Review by majorgreys — COMMENTED** [review]
LGTM with minor questions.

File: tests/debugging/function/test_discovery.py (hunk 1 above)
Comment (P403n1x87) [reply tests/debugging/function/test_discovery.py:48]: 👍 good catch!

**Review by majorgreys — APPROVED** [review]

File: ddtrace/debugging/_function/discovery.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+            )
+            assert container.fget is not None
+            self.__name__ = container.fget.__name__
+
+        elif isinstance(container, (classmethod, staticmethod)):
+            self._iter = iter([(type(container), container.__func__)])  # type: ignore[list-item]
+            self.__name__ = None
+
+        else:
+            raise TypeError("Unsupported container type: %s", type(container))
```
Comment (ZStriker19) [inline ddtrace/debugging/_function/discovery.py:109]: Do we need to catch this somewhere and write a log or we'd break a customer's application? I'm not sure what standard practice is regarding raising errors and catching them from the tracer standpoint, but I could imagine a customer being annoyed that we're keeping their application from running.

**Review by ZStriker19 — APPROVED** [review]
Just the one question about the error the but otherwise looks good!

File: ddtrace/debugging/_function/discovery.py (hunk 2 above)
Comment (P403n1x87) [reply ddtrace/debugging/_function/discovery.py:109]: That's a good point. This error should be handled by the user of the `FunctionDiscovery` class. But, in reality, we should never see this exception being raised. I've put this here just for good measure, but ideally we'd never hit this branch.

**Pull request merged**